### PR TITLE
This PR adds a chart for the newrelic-infra agent.

### PR DESCRIPTION
**Pull request by rk295**
This is largely a copy of `https://github.com/kubernetes/kubernetes/tree/master/examples/newrelic-infrastructure` and configured to be a chart.

### CODE REVIEW

File: stable/newrelic-infrastructure/templates/_helpers.tpl [hunk 1]
```diff
@@ -0,0 +1,30 @@
+{{/* vim: set filetype=mustache: */}}
+{{/*
+Expand the name of the chart.
+*/}}
+{{- define "name" -}}
```
Comment (unguiculus) [inline stable/newrelic-infrastructure/templates/_helpers.tpl]: Should be namespaced as well: `{{- define "newrelic-infra.name" -}}`

File: stable/newrelic-infrastructure/templates/_helpers.tpl [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+{{- else -}}
+{{- printf "%s" $name | trunc 63 | trimSuffix "-" -}}
+{{- end -}}
+{{- end -}}
+
+{{/* Generate basic labels */}}
+{{- define "newrelic-infra.labels" }}
+app: {{ template "name" . }}
+heritage: {{.Release.Service }}
+release: {{.Release.Name }}
```
Comment (unguiculus) [inline stable/newrelic-infrastructure/templates/_helpers.tpl]: Why not include the `chart` label here as well?

File: stable/newrelic-infrastructure/templates/daemonset.yaml [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+          {{- if .Values.resources }}
+          resources:
+{{ toYaml .Values.resources | indent 12 }}
+          {{- end }}
+      volumes:
+        - name: newrelic-config
+          secret:
+            secretName: newrelic-config
+        - name: dev
+          hostPath:
```
Comment (unguiculus) [inline stable/newrelic-infrastructure/templates/daemonset.yaml:68]: Consistently indent with two spaces.

File: stable/newrelic-infrastructure/templates/secret.yaml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: newrelic-config
```
Comment (unguiculus) [inline stable/newrelic-infrastructure/templates/secret.yaml]: Add standard labels.

File: stable/newrelic-infrastructure/templates/_helpers.tpl [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+{{- if .Values.podLabels}}
```
Comment (unguiculus) [inline stable/newrelic-infrastructure/templates/_helpers.tpl]: I would keep pod labels separately because standard labels should go on all resources, not only pod.

File: stable/newrelic-infrastructure/templates/daemonset.yaml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
+  name: {{ template "newrelic-infra.fullname" . }}
+spec:
+  template:
+    metadata:
+    {{- if .Values.podAnnotations }}
+      annotations:
+{{ toYaml .Values.podAnnotations | indent 8}}
+    {{- end }}
+      labels: {{ include "newrelic-infra.labels" . | indent 8 }}
```
Comment (unguiculus) [inline stable/newrelic-infrastructure/templates/daemonset.yaml]: Usually only `app` and `release` labels are added here.

File: stable/newrelic-infrastructure/values.yaml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+# This can be set, the default is shown below
+# logFile: /var/log/nr-infra.log
+
+image:
+  name: newrelic/infrastructure
+  tag: 0.0.7
+  pullPolicy: IfNotPresent
+
+resources:
```
Comment (unguiculus) [inline stable/newrelic-infrastructure/values.yaml]: We usually recommend not to specify defaults for resources leaving this a conscious choice of the user.

**Review by unguiculus — CHANGES_REQUESTED** [review]
Thanks for the PR and sorry for the delay. Will the chart come up without license key? You may want to add some check as is done e. g. here: https://github.com/kubernetes/charts/tree/master/stable/datadog

**Comment by unguiculus** [comment]
Marking this as stale. Please update within one week.

**Comment by rk295** [comment]
Sorry for lag - I'll address the comments above shortly!

File: stable/newrelic-infrastructure/templates/_helpers.tpl (hunk 2 above)
Comment (rk295) [reply stable/newrelic-infrastructure/templates/_helpers.tpl]: I was advised on a previous PR (#1254) that the pods should have three labels but other resources can have four. I only put three here so I can use this helper everywhere, only adding the `chart` label when this isn't applied to a pod, if that makes sense? I was trying to maintain consistency with my other PR.

File: stable/newrelic-infrastructure/templates/_helpers.tpl (hunk 1 above)
Comment (rk295) [reply stable/newrelic-infrastructure/templates/_helpers.tpl]: 👍 fixed

File: stable/newrelic-infrastructure/templates/_helpers.tpl (hunk 5 above)
Comment (rk295) [reply stable/newrelic-infrastructure/templates/_helpers.tpl]: See above. I'm applying the standard labels to all resources, only adding `chart` when it is safe to do so.

File: stable/newrelic-infrastructure/templates/daemonset.yaml (hunk 6 above)
Comment (rk295) [reply stable/newrelic-infrastructure/templates/daemonset.yaml]: as above

File: stable/newrelic-infrastructure/templates/daemonset.yaml (hunk 3 above)
Comment (rk295) [reply stable/newrelic-infrastructure/templates/daemonset.yaml:68]: 👍

File: stable/newrelic-infrastructure/templates/secret.yaml (hunk 4 above)
Comment (rk295) [reply stable/newrelic-infrastructure/templates/secret.yaml]: as above

File: stable/newrelic-infrastructure/values.yaml (hunk 7 above)
Comment (rk295) [reply stable/newrelic-infrastructure/values.yaml]: 👍

Comment (rk295) [reply stable/newrelic-infrastructure/values.yaml]: I'm not sure I agree, if the user knows little about the application then they might not know what is a sane default. In this case these defaults come from the vendor of the application, so can be trusted (imho).  I looked through all the stable charts before making this choice 52 specify resources with only 18 leaving them empty. 

Happy to remove if its a requirement of entry but I wanted to explain my thinking here!

**Comment by rk295** [comment]
@unguiculus I've pushed some changes and some further comments above.

File: stable/newrelic-infrastructure/templates/_helpers.tpl (hunk 2 above)
Comment (unguiculus) [reply stable/newrelic-infrastructure/templates/_helpers.tpl]: This is how it's usually done:
* All four labels on resources
* Labels `app` and `release` only on pods

File: stable/newrelic-infrastructure/values.yaml (hunk 7 above)
Comment (unguiculus) [reply stable/newrelic-infrastructure/values.yaml]: The reason we now tend to recommend to not specify defaults is that users should be able to install/test charts on small systems, such as Minikube. Oftentimes, defaults were kind of really high. In your case, I guess you can leave them in.

File: stable/newrelic-infrastructure/templates/daemonset.yaml (hunk 3 above)
Comment (unguiculus) [reply stable/newrelic-infrastructure/templates/daemonset.yaml:68]: Make list indentation consistent as well, i. e. content below `env` should be indented one more level.

Comment (rk295) [reply stable/newrelic-infrastructure/templates/daemonset.yaml:68]: 👍

File: stable/newrelic-infrastructure/templates/_helpers.tpl (hunk 5 above)
Comment (unguiculus) [reply stable/newrelic-infrastructure/templates/_helpers.tpl]: Sorry to be picky. Labels `chart` and `heritage` should not go on pods.

See common chart for reference of a label template: https://github.com/kubernetes/charts/blob/HEAD/incubator/common/templates/_metadata_labels.tpl#L21-L27

Also see https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md. 

Or look at the initial chart you get with `helm create`.

**Comment by unguiculus** [comment]
You branch is way behind master. Please update it.

File: stable/newrelic-infrastructure/templates/_helpers.tpl (hunk 5 above)
Comment (rk295) [reply stable/newrelic-infrastructure/templates/_helpers.tpl]: fixed as per your advice.

File: stable/newrelic-infrastructure/templates/daemonset.yaml (hunk 3 above)
Comment (rk295) [reply stable/newrelic-infrastructure/templates/daemonset.yaml:68]: Also fixed now.

**Comment by rk295** [comment]
@unguiculus master now merged in, you should find this branch is up to date now.

**Comment by unguiculus** [comment]
Please fix linting issues.

**Comment by rk295** [comment]
I've fixed them from what I can tell, but with helm 2.7.2 I show now errors, or is the linting done by tiller ?

```
% helm lint stable/newrelic-infrastructure/
==> Linting stable/newrelic-infrastructure/
Lint OK

1 chart(s) linted, no failures
% helm version
Client: &version.Version{SemVer:"v2.7.2", GitCommit:"8478fb4fc723885b155c924d1c8c410b7a9444e6", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.4.2", GitCommit:"82d8e9498d96535cc6787a6a9194a76161d29b4c", GitTreeState:"clean"}
```

**Comment by unguiculus** [comment]
We do some additional linting. See https://github.com/kubernetes/charts/blob/master/test/circle/lint.sh

**Comment by rk295** [comment]
@unguiculus The lint is passing now. I don't understand the Jenkins failure. Firstly it seems to be just detecting the string `Error` - which is shown from `NOTES.txt` if no license key is provided. That is the intended behaviour, happy to change to a warning.  And the following lines don't seem to be related to this chart:

```
E1209 20:27:11.740] Build timed out
E1209 20:27:11.741] Command failed
I1209 20:27:11.741] process 6818 exited with code -15 after 10.0m
E1209 20:27:11.741] FAIL: pull-charts-e2e
```

**Comment by rk295** [comment]
@unguiculus do you want me to remove word `ERROR` from the `NOTES.txt` so it passes the Jenkins job? I think I've addressed all your other comments now.

File: stable/newrelic-infrastructure/values.yaml [hunk 8]
```diff
@@ -0,0 +1,64 @@
+# IMPORTANT: Specify your New Relic API key here.
+# licenseKey:
+
+verboseLog: false
+
+# This can be set, the default is shown below
+# logFile: /var/log/nr-infra.log
+
+image:
+  name: newrelic/infrastructure
```
Comment (unguiculus) [inline stable/newrelic-infrastructure/values.yaml]: `name` -> `repository`

File: stable/newrelic-infrastructure/.gitignore [hunk 9]
```diff
@@ -0,0 +1 @@
+*.local
```
Comment (unguiculus) [inline stable/newrelic-infrastructure/.gitignore]: Why is this needed?

File: stable/newrelic-infrastructure/Chart.yaml [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+name: newrelic-infrastructure
+version: 0.0.1
+appVersion: 0.0.12
+home: https://hub.docker.com/r/newrelic/infrastructure/
+source:
+  - https://github.com/kubernetes/kubernetes/tree/master/examples/newrelic-infrastructure
+engine: gotpl
+icon: https://newrelic.com/assets/newrelic/source/NewRelic-logo-square.svg
+maintainers:
+  - name: Robin Kearney
```
Comment (unguiculus) [inline stable/newrelic-infrastructure/Chart.yaml]: Use Github username

**Review by unguiculus — CHANGES_REQUESTED** [review]
The problem is not the ERROR vs WARNING thing. The pods never come up. I would guess this is because you create the daemon set even if no licence is specified.

File: stable/newrelic-infrastructure/.gitignore (hunk 9 above)
Comment (rk295) [reply stable/newrelic-infrastructure/.gitignore]: My mistake, gone.

File: stable/newrelic-infrastructure/Chart.yaml (hunk 10 above)
Comment (rk295) [reply stable/newrelic-infrastructure/Chart.yaml]: fixed.

**Comment by rk295** [comment]
@unguiculus I fixed up the CI tests and your comments above.
